### PR TITLE
DCR: Strongly Typed Activities

### PR DESCRIPTION
**Pull request by drub0y**
This represents the work related to #761.

## Open Issues

- [x] Figure out why "Classic" projects are not building with C# 7 (currently erroring on pattern matching `switch` statements)
- [ ] Coming to grips with the fact that `Microsoft.Bot.Schema` is no longer able to be fully code generated
- [ ] All samples need to be upgraded to reflect these API changes

### CODE REVIEW

**Comment by drub0y** [comment]
So in the state it _is_ in right now (pre-this PR), I would agree with you; it would be much better to just remove those concepts and just have the uber `Activity` class. In the state that I'm aiming for in this DCR, where you can actually rely on the types and are actually "forced" to work with the correct types in order to access the correct properties, I actually think it's much better and would disagree with reverting to a single uber class.

I did go through all the code that's out there (SDK, not samples yet) and change it to work against these new types and it reduced the noise quite a bit and gave me what I expect as a .NET developer. I think the impact will be greater the further you get out from the core though: in middleware, the dialogs layer and bots themselves. That said, there were some type specific checks such as the adapters checking for `InvokeResponseActivity`, `DelayActivity`, `TraceActivity`, etc. that became much cleaner (IMHO) under this design.

I'm still going to do another pass over the code to make sure there's no more `activity.Type == ` checks and, if there are any left, I'll make convert those to strongly typed.

**Comment by drub0y** [comment]
Rebased against latest from `master`.

File: libraries/Microsoft.Bot.Schema/Activity.cs [hunk 1]
```diff
@@ -1,14 +1,4 @@
-// <auto-generated>
-// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See License.txt in the project root for
-// license information.
-//
-// Code generated by Microsoft (R) AutoRest Code Generator.
-// Changes may cause incorrect behavior and will be lost if the code is
-// regenerated.
-// </auto-generated>
-
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Schema/Activity.cs]: I totally get why you want to go this way, but I think this is a non-starter.  We are constantly updating the schema for new types, and have to keep the changes in sync with multiple languages. 
I just don't see that we have the bandwidth to make this 100% hand coded.  This is a huge delta, and I wish you had talked with me before embarking on it.

**Comment by drub0y** [comment]
Rebased against latest (state management again) in `master`.

**Comment by johnataylor** [comment]
I agree strongly with @cleemullins on this one. The "strong" types don't add much value - they are after all protocol messages rather than application logic. For example, we only have one class for HTTP requests independent of the method. Its also pretty clear from the existing code that the typing adds little value.

**Comment by tomlm** [comment]
We are not going to take this change. I am closing this PR as it is getting very out of date.

**Pull request closed without merging**